utils/clflags.ml

```diff
@@ -397,53 +399,123 @@ module Extension = struct
+  let is_erasable = function
+    | Local ->
+        true
+    | Comprehensions
+    | Include_functor
+    | Polymorphic_parameters
+    | Immutable_arrays ->
+        false
+
+  module Universe = struct
+    (** Which extensions can be enabled? *)
+    type t =
+      | No_extensions
+      | Only_erasable
+      | Any
+
+    let compiler_options = function
+      | No_extensions -> "flag -disable-all-extensions"
+      | Only_erasable -> "flag -only-erasable-extensions"
+      | Any           -> "default options"
+
+    let is_allowed t ext = match t with
+      | No_extensions -> false
+      | Only_erasable -> is_erasable ext
+      | Any           -> true
+  end
+
+  (* Mutable state.  Invariants:
 
-  let enable extn =
-    if !disable_all_extensions then
+     (1) [!extensions] contains at most one copy of each extension.
+
+     (2) Every member of [!extensions] satisfies [Universe.is_allowed !universe].
+         (For instance, [!universe = No_extensions] implies
+         [!extensions = []]). *)
+  let extensions = ref default_extensions (* -extension *)
+  let universe   = ref Universe.Any       (* -only-erasable-extensions,
+                                             -disable-all-extensions *)
+
+  let set extn ~enabled =
+    if not (Universe.is_allowed !universe extn) then
       raise (Arg.Bad(Printf.sprintf
-        "Cannot enable extension %s: \
-         incompatible with compiler flag -disable-all-extensions"
-        extn));
-    match of_string (String.lowercase_ascii extn) with
-    | Some extension ->
-        if not (List.exists (equal extension) !extensions) then
-          extensions := extension :: !extensions
-    | None ->
-        raise (Arg.Bad (Printf.sprintf "Unknown extension \"%s\"" extn))
-
-  let is_enabled ext =
-    not !disable_all_extensions
-    && (List.mem ext default_extensions
-        || List.mem ext !extensions)
+        "Cannot %s extension %s: incompatible with %s"
+        (if enabled then "enable" else "disable")
+        (to_string extn)
+        (Universe.compiler_options !universe)));
+    if enabled then begin
+      if not (List.exists (equal extn) !extensions) then
+        extensions := extn :: !extensions
+    end else
+      (* Like [filter (( <> ) extn)], but stop after deleting the first one *)
+      let rec del = function
+        | [] -> []
+        | extn' :: extns when equal extn extn' -> extns
+        | extn' :: extns -> extn' :: del extns
+      in
+      extensions := del !extensions
+
+  let enable  = set ~enabled:true
+  let disable = set ~enabled:false
+
+  let is_enabled extn = List.mem extn !extensions
+
+  (* It might make sense to ban [set], [enable], [disable],
+     [only_erasable_extensions], and [disallow_extensions] inside [f], but it's
+     not clear that it's worth the hassle *)
+  let with_set extn ~enabled f =
+    let current_extensions = !extensions in
+    let current_universe   = !universe   in
+    Fun.protect
```

Comment:
I think this function basically just inlines `Misc.protect_refs`. Maybe just use that instead?

Reply:
Not quite?  `Misc.protect_refs` sets them directly, which we don't want to do here.